contracts/extensions/ERC4907A.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: MIT
+// ERC721A Contracts v4.1.0
+// Creator: Chiru Labs
+
+pragma solidity ^0.8.4;
+
+import './IERC4907A.sol';
+import '../ERC721A.sol';
+
+/**
+ * @dev ERC4907 compliant extension of ERC721A.
+ *
+ * The ERC4907 standard https://eips.ethereum.org/EIPS/eip-4907[ERC4907] allows
+ * owners and authorized addresses to add a time-limited role
+ * with restricted permissions to ERC721 tokens.
+ */
+abstract contract ERC4907A is ERC721A, IERC4907A {
+    // The bit position of `expires` in packed user info.
+    uint256 private constant _BITPOS_EXPIRES = 160;
+
+    // The mask of the lower 160 bits for addresses.
+    uint256 private constant _BITMASK_ADDRESS = (1 << 160) - 1;
+
+    // Mapping from token ID to user info.
+    //
+    // Bits Layout:
+    // - [0..159]   `user`
+    // - [160..223] `expires`
+    mapping(uint256 => uint256) private _packedUserInfo;
+
+    /**
+     * @dev Sets the `user` and `expires` for `tokenId`.
+     * The zero address indicates there is no user.
+     *
+     * Requirements:
+     *
+     * - The caller must own `tokenId` or be an approved operator.
+     */
+    function setUser(
+        uint256 tokenId,
+        address user,
+        uint64 expires
+    ) public {
+        if (!_isApprovedOrOwner(msg.sender, tokenId)) revert SetUserCallerNotOwnerNorApproved();
+
+        _packedUserInfo[tokenId] = (uint256(expires) << _BITPOS_EXPIRES) | uint256(uint160(user));
+
+        emit UpdateUser(tokenId, user, expires);
+    }
+
+    /**
+     * @dev Returns the user address for `tokenId`.
+     * The zero address indicates that there is no user or if the user is expired.
+     */
+    function userOf(uint256 tokenId) public view returns (address) {
+        uint256 packed = _packedUserInfo[tokenId];
+        assembly {
+            // Branchless `packed *= !(block.timestamp > expires) ? 1 : 0`.
+            packed := mul(
+                packed,
+                // `!(block.timestamp > expires) ? 1 : 0`.
+                iszero(
+                    gt(
+                        timestamp(),
+                        shr(_BITPOS_EXPIRES, packed) // `expires`.
+                    )
+                )
+            )
+        }
+        return address(uint160(packed));
```

Comment:
What do you think about dropping the negation ?
```solidity 
      function userOf(uint256 tokenId) public view returns (address) {
         uint256 packed = _packedUserInfo[tokenId];
         assembly {
             // Branchless `packed *= expires > block.timestamp  ? 1 : 0`.
             packed := mul(
                 packed,
                 // `expires > block.timestamp ? 1 : 0`.
                 gt(
                     shr(_BITPOS_EXPIRES, packed), // `expires`.
                     timestamp()
                 )
             )
         }
         return address(uint160(packed));
```

Reply:
Lemme think hard about it. 

The negation is to be faithful to the reference implementation.

Reply:
Ok, i did something. If you know you know.

Reply:
I did a little trick. ;)

Now it’s faithful to the reference.

Reply:
**Explanation**

First, we know:

`packed = (expires << 160) | user`

The straightforward way of implementing `block.timestamp <= expires` in is:

`!(block.timestamp > (packed >> 160))` which in Yul, is: `iszero(gt(timestamp(), shr(160, packed)))`


Now, notice that if `user != address(0)`:

`(block.timestamp << 160) < (expires << 160) | user` holds true for `block.timestamp <= expires`.

And when `user == address(0)`:

`address(uint160((expires << 160) | user))` will give `address(0)` anyway.

So that's how we remove the `iszero`.